### PR TITLE
Add HD OSD support

### PR DESCRIPTION
**Pull request by SteveCEvans**
Adds support for HD OSD, standardising on a 53x20 grid, utilising 24x36 fonts for 720P and 36x54 fonts for 1080P.

Grids for NTSC, PAL and HD are now supported.

image

And whilst moving items the grid is now a checkerboard.

image

This is for discussion. Coordinating setting of `vcd_video_system` and `osd_displayport_device` and improving the font (ideally colour to match the HD goggles in HD mode) are clear obvious enhancements.

See also https://github.com/betaflight/betaflight/pull/11964.

### CODE REVIEW

**Comment by McGiverGim** [comment]
I think I prefer the dashed background (like here https://github.com/betaflight/betaflight-configurator/pull/3008 and not the checked one. Other thoughts?

**Comment by SteveCEvans** [comment]
It’s just an image (or rather a set of three). I think the old dashed one looked a bit clunky, hence this suggestion. Happy to use whatever image people find clearest.

**Comment by McGiverGim** [comment]
The link I posted removes the background image and uses css to draw it.

**Comment by blckmn** [comment]
AUTOMERGE: (FAIL)
 - github identifies PR as mergeable -> FAIL
 - assigned to a milestone -> FAIL
 - cooling off period lapsed -> FAIL
 - commit count less or equal to three -> PASS
 - `Don't merge` label NOT found -> PASS
 - at least one `RN:` label found -> FAIL
 - `Tested` label found -> FAIL
 - assigned to an approver -> FAIL
 - approver count at least three -> FAIL

**Comment by SteveCEvans** [comment]
@McGiverGim Yes, I appreciate that. Both PRs take the obvious step of defining `VIDEO_COLS` to define the screen size. The more significant change in this PR is to handle the 6 bit X encoding. I’m open to different CSS.

**Comment by McGiverGim** [comment]
Please verify the unit tests. Is detecting position declared twice.

File: src/js/tabs/osd.js
Comment (haslinghuis) [inline src/js/tabs/osd.js:651]: This line overwrites previous initialization

File: src/js/tabs/osd.js
Comment (haslinghuis) [inline src/js/tabs/osd.js:685]: Probably should be
```suggestion
                return (OSD.constants.VIDEO_COLS[videoType] >> 1) + ((OSD.constants.VIDEO_ROWS[videoType] >> 1) - 2) * OSD.constants.VIDEO_COLS[videoType] - 1;
```
or define on line 691 and return position.

File: src/js/tabs/osd.js
Comment (haslinghuis) [inline src/js/tabs/osd.js:1979]: Indentation

File: src/js/tabs/osd.js
Comment (SteveCEvans) [reply src/js/tabs/osd.js]: Spotted that. Thanks.

File: src/js/tabs/osd.js
Comment (SteveCEvans) [reply src/js/tabs/osd.js:686]: Agreed. There were three such instances left over from a code simplification. Fixed.

File: src/js/tabs/osd.js
Comment (SteveCEvans) [reply src/js/tabs/osd.js:1980]: No longer needed, but thanks.

**Comment by benlumley** [comment]
If anyone wants me to merge the CSS change + ditch the images as @McGiverGim is prefering; @me and i'll happily do it on top of this branch. Pretty sure it's an easy copy/paste mind!

And to add - my thinking for switching to CSS was to remove the requirement to create new images for any other new screen sizes that may come into being.

**Review by McGiverGim — APPROVED** [review]
Seems good to me!

**Comment by haslinghuis** [comment]
My board freaked out after saving to HD on osd tab and going to motors tab.

image

After several manual connection attempts it went into DFU mode.
Something in d3 library is not happy. It's repeatable but sometimes it works.

File: src/js/serial_backend.js
Comment (haslinghuis) [inline src/js/serial_backend.js:729]: Please move to OSD tab

Comment (SteveCEvans) [reply src/js/serial_backend.js]: Fixed

File: src/js/tabs/osd.js
Comment (McGiverGim) [inline src/js/tabs/osd.js:1967]: One question. We have a way to know if there is an HD system configured/attached? I don't have the device, so I'm not sure if we have this information. In this case it will be interesting to select HD and not PAL here.

**Review by McGiverGim — APPROVED** [review]
Approved, but one question...

File: src/js/tabs/osd.js
Comment (SteveCEvans) [reply src/js/tabs/osd.js:1967]: What I could do is change the firmware so that if a `MSP_SET_OSD_CANVAS` command is received, the `vcd_video_system` could be set to `HD`. If a save is done subsequent to that with the VTX connected, then both that and the canvas size would be preserved and used even if the VTX isn't powered.

See https://github.com/betaflight/betaflight/pull/11964/commits/fa11425550bc385c7959c1a056b5b6f7e6703944

Comment (McGiverGim) [reply src/js/tabs/osd.js:1967]: Is not enough if MSP+Displayport is selected? Is simply a best effort, we let the user modify it if needed.

**Review by haslinghuis — DISMISSED** [review]

File: src/js/tabs/osd.js
Comment (SteveCEvans) [reply src/js/tabs/osd.js:1967]: MSP VTX and displayport are one combined option and don’t necessarily mean HD.

**Review by sugaarK — APPROVED** [review]

**Comment by SteveCEvans** [comment]
I've now added the system elements so they can be positioned using the configurator.

Ideally these should be rendered using a font that's more representative than the MAX7456 fonts, so `Mbps` can be shown correctly, for example.

image

**Comment by haslinghuis** [comment]
OsdTextElementSysBitrate

**Comment by SteveCEvans** [comment]
> OsdTextElementSysBitrate

Good spot! I've had a bit of a think about the semantics of the element names and revised them in addition to fixing this.

File: src/js/tabs/osd.js
Comment (haslinghuis) [inline src/js/tabs/osd.js:1438]: Please add ending comma

File: src/js/tabs/osd.js
Comment (haslinghuis) [inline src/js/tabs/osd.js:1900]: Please use space (instead of tabs)

Comment (SteveCEvans) [reply src/js/tabs/osd.js]: Done

File: src/js/tabs/osd.js
Comment (SteveCEvans) [reply src/js/tabs/osd.js]: Done

File: src/js/tabs/osd.js
Comment (haslinghuis) [inline src/js/tabs/osd.js:1420]: Sorry sonar is really picky: 

```suggestion
            preview: `10${FONT.symbol(SYM.METRE)}`,
```

Same on  
/home/runner/work/betaflight-configurator/betaflight-configurator/src/js/tabs/osd.js
Error:   1420:22  error  Unexpected string concatenation  prefer-template
Error:   1429:22  error  Unexpected string concatenation  prefer-template
Error:   1438:22  error  Unexpected string concatenation  prefer-template
Error:   1447:22  error  Unexpected string concatenation  prefer-template

Comment (SteveCEvans) [reply src/js/tabs/osd.js]: Yup, spotted that!

Comment (haslinghuis) [reply src/js/tabs/osd.js]: Thanks. It's building now!!!

Comment (haslinghuis) [reply src/js/tabs/osd.js]: @SteveCEvans guess you didn't notice the back ticks here otherwise the ${var} template literal is not expanded

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Template_literals

File: src/js/tabs/osd.js
Comment (haslinghuis) [inline src/js/tabs/osd.js:1420]:
```suggestion
            preview: `10${FONT.symbol(SYM.METRE)}`,
```

File: src/js/tabs/osd.js
Comment (haslinghuis) [inline src/js/tabs/osd.js:1429]:
```suggestion
            preview: `G${FONT.symbol(SYM.LINK_QUALITY)}100`,
```

File: src/js/tabs/osd.js
Comment (haslinghuis) [inline src/js/tabs/osd.js:1438]:
```suggestion
            preview: `${FONT.symbol(SYM.ARROW_SMALL_RIGHT)}G DVR 8.4G`,
```

File: src/js/tabs/osd.js
Comment (haslinghuis) [inline src/js/tabs/osd.js:1447]:
```suggestion
            preview: `${FONT.symbol(SYM.ARROW_SMALL_RIGHT)}A DVR 1.6G`,
```

**Comment by haslinghuis** [comment]
@SteveCEvans guess you didn't notice the back ticks here otherwise the ${var} template literal is not expanded

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Template_literals

File: src/js/tabs/osd.js
Comment (SteveCEvans) [reply src/js/tabs/osd.js:1447]: But they are... 
image

Happy to change anyway for consistency.

Comment (SteveCEvans) [reply src/js/tabs/osd.js]: Fixed

**Review by haslinghuis — APPROVED** [review]

File: src/js/tabs/osd.js
Comment (haslinghuis) [inline src/js/tabs/osd.js:1888]:
```suggestion
                                                            if (semver.gte(FC.CONFIG.apiVersion, API_VERSION_1_45)) {
```

**Review by haslinghuis — CHANGES_REQUESTED** [review]

**Comment by sugaarK** [comment]
Will review once changes made.. really happy this got bumped to 10.9 this is a big feature that will make the new release more exciting for the people

File: src/js/tabs/osd.js
Comment (SteveCEvans) [reply src/js/tabs/osd.js:1888]: But the current FW is 1.46 MSP which is there the F.SYS stuff got added.

Comment (haslinghuis) [reply src/js/tabs/osd.js:1888]: We should target this for 1.45

File: src/js/tabs/osd.js
Comment (haslinghuis) [inline src/js/tabs/osd.js:1476]: Missing trailing comma

File: src/js/data_storage.js
Comment (KarateBrot) [inline src/js/data_storage.js:18]:
```suggestion
```

File: src/js/tabs/osd.js
Comment (KarateBrot) [inline src/js/tabs/osd.js:2676]:
```suggestion
    if (semver.gte(FC.CONFIG.apiVersion, API_VERSION_1_45)) {
```

**Review by KarateBrot — CHANGES_REQUESTED** [review]

File: src/js/msp/MSPHelper.js
Comment (McGiverGim) [inline src/js/msp/MSPHelper.js:1701]: Not important, but the rest of cases not include an empty line before the break, so better to maintain the coherence in the style.
This is a read MSP message? In theory all the "read" commands go higher, is easy to see them because they contain code, and we have here, at bottom only "SET" commands, that usually only write to the console.

File: src/js/data_storage.js
Comment (SteveCEvans) [reply src/js/data_storage.js:18]: Gotcha. Now that the FW API version has reverted to 1.45 this is the thing to do.
Fixed

File: src/js/msp/MSPHelper.js
Comment (SteveCEvans) [reply src/js/msp/MSPHelper.js:1701]: Blank line removed.

**Comment by sugaarK** [comment]
@SteveCEvans  what’s the B concuss on the @KarateBrot  changes ?

**Review by haslinghuis — APPROVED** [review]

**Review by KarateBrot — APPROVED** [review]

**Pull request merged**
